### DISCUSSION TITLE
MRENCLAVE VS .sig remote attestation

### DISCUSSION THREAD

**Discussion by nmwael** [gramineproject/gramine#1357] [Q&A] [answered]
Hi 

Are there anyway to verify that a running enclave (MRENCLAVE ) are derived from a certain gramized application .sig? 

Or are there another way of checking this? For example using isv_prod_id and isv_svn ?

The question was spawned from this:
https://github.com/gramineproject/gramine/issues/1343#issuecomment-1550851681

**Comment by dimakuv** [comment, marked as answer]
> Are there anyway to verify that a running enclave (MRENCLAVE ) are derived from a certain gramized application .sig?

Yes.

Gramine dumps the SIGSTRUCT object into the `.sig` file. This SIGSTRUCT object has a bunch of Intel-SGX-technology-standardized fields. Here you can take a look at these fields: https://github.com/gramineproject/gramine/blob/ede508c69217477c1cd6fdb3e7689da824ba4ea7/python/graminelibos/sigstruct.py#L19-L41

One of these fields is called `enclave_hash` in Gramine code, but it is typically named MRENCLAVE.

So you can extract the MRENCLAVE from the `.sig` file by reading bytes from a specific offset in this file. Or by using the Gramine Python API to access this SIGSTRUCT object. Or by using the `gramine-sgx-view-sigstruct` tool that will appear in the next release of Gramine.

---

Now can you extract MRENCLAVE from the SGX quote? Yes, you can. This value is found in `sgx_quote::sgx_report::mrenclave` field.

So if you have the SIGSTRUCT file and you received the SGX quote from a remote SGX enclave, you can verify that this SGX enclave has an expected MRENCLAVE by comparing `sigstruct::enclave_hash == sgx_quote::sgx_report::mrenclave`.

**Reply by nmwael** [reply]
Thanks a lot.
